File: apps/api/app/schemas/capital_ledger.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, field_serializer


CapitalPoolStatus = Literal["active", "inactive", "completed", "cancelled"]
CapitalPoolType = Literal[
    "paper_account",
    "validation_run",
    "research_campaign",
    "strategy_allocation",
    "position",
    "compounding_recommendation",
    "withdrawal_recommendation",
    "profit_reserve",
    "policy_review",
]


class CapitalLedgerSummaryResponse(BaseModel):
    total_managed_capital: Decimal
    total_starting_capital: Decimal
    total_current_equity: Decimal
    total_allocated_capital: Decimal
    total_available_capital: Decimal
    total_reserved_capital: Decimal
    total_realized_pnl: Decimal
    total_unrealized_pnl: Decimal
    active_capital_pools: int
    inactive_capital_pools: int
    active_positions: int
    total_trades: int
    utilization_percent: float
    data_completeness_percent: float
    unavailable_sources: list[str]
    generated_at: datetime

    @field_serializer(
        "total_managed_capital",
        "total_starting_capital",
        "total_current_equity",
        "total_allocated_capital",
        "total_available_capital",
        "total_reserved_capital",
        "total_realized_pnl",
        "total_unrealized_pnl",
        when_used="json",
    )
    def serialize_decimal_fields(self, value: Decimal) -> str:
        return format(value, "f")


class CapitalPoolResponse(BaseModel):
    capital_pool_id: str
    capital_pool_type: CapitalPoolType
    name: str
    status: CapitalPoolStatus
    starting_capital: Decimal | None
    current_equity: Decimal | None
    allocated_capital: Decimal | None
    available_capital: Decimal | None
    reserved_capital: Decimal | None
    realized_pnl: Decimal | None
    unrealized_pnl: Decimal | None
    pnl_percent: float | None
    started_at: datetime | None
    completed_at: datetime | None
    related_entity_type: str
    related_entity_id: str
    related_page_url: str
    capital_campaign_uuid: str | None = None
    capital_campaign_name: str | None = None
    capital_campaign_status: str | None = None
    parent_capital_pool_id: str | None = None
    child_allocations_count: int
    notes: str | None = None

    @field_serializer(
        "starting_capital",
        "current_equity",
        "allocated_capital",
        "available_capital",
        "reserved_capital",
        "realized_pnl",
        "unrealized_pnl",
        when_used="json",
    )
    def serialize_optional_decimal_fields(self, value: Decimal | None) -> str | None:
        if value is None:
            return None
        return format(value, "f")


class CapitalLedgerResponse(BaseModel):
    summary: CapitalLedgerSummaryResponse
    capital_pools: list[CapitalPoolResponse]
    page: int
    page_size: int
    total: int
    has_more: bool

File: apps/web/lib/api/capital-ledger.ts
import { ApiRequestError } from "@/lib/api/arena";

export type CapitalLedgerStatus = "all" | "active" | "inactive" | "completed" | "cancelled";
export type CapitalLedgerType =
  | "all"
  | "paper_account"
  | "validation_run"
  | "research_campaign"
  | "strategy_allocation"
  | "position"
  | "compounding_recommendation"
  | "withdrawal_recommendation"
  | "profit_reserve"
  | "policy_review";

export type CapitalLedgerSummary = {
  total_managed_capital: string;
  total_starting_capital: string;
  total_current_equity: string;
  total_allocated_capital: string;
  total_available_capital: string;
  total_reserved_capital: string;
  total_realized_pnl: string;
  total_unrealized_pnl: string;
  active_capital_pools: number;
  inactive_capital_pools: number;
  active_positions: number;
  total_trades: number;
  utilization_percent: number;
  data_completeness_percent: number;
  unavailable_sources: string[];
  generated_at: string;
};

export type CapitalLedgerPool = {
  capital_pool_id: string;
  capital_pool_type: CapitalLedgerType;
  name: string;
  status: Exclude<CapitalLedgerStatus, "all">;
  starting_capital: string | null;
  current_equity: string | null;
  allocated_capital: string | null;
  available_capital: string | null;
  reserved_capital: string | null;
  realized_pnl: string | null;
  unrealized_pnl: string | null;
  pnl_percent: number | null;
  started_at: string | null;
  completed_at: string | null;
  related_entity_type: string;
  related_entity_id: string;
  related_page_url: string;
  capital_campaign_uuid?: string | null;
  capital_campaign_name?: string | null;
  capital_campaign_status?: string | null;
  parent_capital_pool_id?: string | null;
  child_allocations_count: number;
  notes?: string | null;
};

export type CapitalLedgerResponse = {
  summary: CapitalLedgerSummary;
  capital_pools: CapitalLedgerPool[];
  page: number;
  page_size: number;
  total: number;
  has_more: boolean;
};

type ErrorEnvelope = {
  error?: {
    message?: string;
  };
};

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

async function requestJson<T>(path: string): Promise<T> {
  const response = await fetch(`${API_BASE_URL}${path}`, {
    cache: "no-store",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) {
    let message = `Request failed with status ${response.status}`;
    try {
      const payload = (await response.json()) as ErrorEnvelope;
      if (payload.error?.message) {
        message = payload.error.message;
      }
    } catch {
      // Keep fallback message.
    }
    throw new ApiRequestError(message, response.status);
  }

  return (await response.json()) as T;
}

export async function getCapitalLedger(params?: {
  status?: CapitalLedgerStatus;
  type?: CapitalLedgerType;
  page?: number;
  pageSize?: number;
}): Promise<CapitalLedgerResponse> {
  const query = new URLSearchParams();
  query.set("status", params?.status ?? "all");
  query.set("type", params?.type ?? "all");
  query.set("page", String(params?.page ?? 1));
  query.set("page_size", String(params?.pageSize ?? 50));
  return requestJson<CapitalLedgerResponse>(`/capital/ledger?${query.toString()}`);
}
